keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1462,225 +1278,309 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
+    //
+    // +optional
+    AdminAccess *bool
 
-    // DriverRequests describes all resources that are needed for the
-    // allocated claim. A single claim may use resources coming from
-    // different drivers. For each driver, this array has at most one
-    // entry which then may have one or more per-driver requests.
+    // The amount field can be used to ask for all devices which match the
+    // requirements or a certain fixed number. The default is exactly one device.
     //
-    // May be empty, in which case the claim can always be allocated.
-    DriverRequests []DriverRequests
-}
+    // Note that the overall number of allocated devices per claim is limited
+    // to 32, so asking for one device in more than 32 requests and asking for
+    // more than 32 identical devices in a single request are both not allowed.
+    // This gets checked during validation. The number of devices matching
+    // the requirements is not known in advance, so asking for "all devices"
+    // can fail during allocation.
+    //
+    // +optional
+    Amount *Amount
 
-// DriverRequests describes all resources that are needed from one particular driver.
-type DriverRequests struct {
-    // DriverName is the name used by the DRA driver kubelet plugin.
-    DriverName string
+    // Requirements describe additional contraints that all must be met by a device
+    // to satisfy the request.
+    //
+    // +optional
+    // +listType=atomic
+    Requirements []Requirement
+}
 
-    // VendorParameters are arbitrary setup parameters for all requests of the
-    // claim. They are ignored while allocating the claim.
-    VendorParameters runtime.Object
+// Exactly one field must be set.
+type Amount struct {
+    // All, if set, asks for all devices matching the requirements.
+    // Allocation fails if not all of them are available, unless admin
+    // access is requested. Admin access is granted also for
+    // devices which are in use.
+    //
+    // +optional
+    All bool
 
-    // Requests describes all resources that are needed from the driver.
-    Requests []ResourceRequest
+    // A fixed number of devices which match the same requirements.
+    //
+    // +optional
+    Count *int64
 }
+```
 
-// ResourceRequest is a request for resources from one particular driver.
-type ResourceRequest struct {
-    // VendorParameters are arbitrary setup parameters for the requested
-    // resource. They are ignored while allocating a claim.
-    VendorParameters runtime.Object
+The `DeviceClassName` field may be left empty. The request itself can specify which
+driver needs to provide devices. This leads to some corner cases:
+- Empty `claim.requests`: this is a "null request" which can be satisfied without
+  allocating any device.
+- Empty `claim.requests[].deviceClassName` and empty `requirements`: this is a request
+  which can be satisfied by any device. This does not really make sense, but is not
+  disallowed either.
```

Comment:
I think we should treat it as an error instead of allocating some arbitrary device. That probably wasn't the intent.

Reply:
Changed.